### PR TITLE
Add caching abstraction around PSR-6, PSR-16 and Doctrine Cache

### PR DESCRIPTION
**Pull request by imjoehaines**
## Goal

This PR adds a caching abstraction which can wrap a PSR-6, PSR-16 or Doctrine Cache instance in a unified interface. More abstractions are easy to add by adding a wrapper class and a condition to the `CacheFactory`

This isn't used yet, but will be in an upcoming PR

## Tests

As PSR-6 and PSR-16 are only interfaces, I've added bare-minimum implementations so that we can write tests for them. These should eventually be expanded upon with integration tests against, e.g. a real Memcache instance, but that would require adding even more dev dependencies which I'd like to avoid. We can look into adding tests against an example project in another PR

The implementations are forced to behave the same via the `AdapterTest` abstract test case, which runs the same tests against all three implementations

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/Cache/Adapter/DoctrineCacheAdapter.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            if ($value === false) {
+                // Doctrine Cache specifies returning false if a key is not
+                // found. However, 'false' is a possible value to cache so we
+                // need to make sure the key doesn't exist
+                // TODO we probably don't need to store 'false' in the cache so
+                //      can probably get away without doing this — this would
+                //      save a second round trip to the cache
+                if ($this->cache->contains($key)) {
+                    return false;
+                }
```
Comment (imjoehaines) [inline src/Cache/Adapter/DoctrineCacheAdapter.php:35]: I'm 99.9% sure we don't need this, but I figured I'd leave it until we're certain — happy to remove it now though

**Comment by GrahamCampbell** [comment]
Note that some versions of Laravel don't implement the cache interface correctly. ;)

**Comment by GrahamCampbell** [comment]
https://github.com/GrahamCampbell/Laravel-GitHub/blob/8.8/src/Http/Psr16Cache.php is a PSR-16 adapter for Laravel 5.5+.

**Comment by imjoehaines** [comment]
> Note that some (older) versions of Laravel don't implement the cache interface correctly. ;)

Yikes, looks like it's just the TTL that would affect us though, which isn't super important because we should be using the cached values and resetting them way before hitting the TTL. I'm tempted to leave this as-is and wrap the cache (if necessary) in bugsnag-laravel

File: tests/TestCase.php [hunk 2]
```diff
@@ -28,6 +28,24 @@ public function expectedException($class, $message = null)
         }
     }
 
+    protected function isPhpUnit7()
```
Comment (GrahamCampbell) [inline tests/TestCase.php:31]: this can be static

File: tests/TestCase.php [hunk 3]
```diff
@@ -28,6 +28,24 @@ public function expectedException($class, $message = null)
...
+    {
+        return version_compare($this->phpUnitVersion(), '7.0.0', '>=')
+            && version_compare($this->phpUnitVersion(), '8.0.0', '<');
+    }
+
+    protected function isPhpUnit6()
```
Comment (GrahamCampbell) [inline tests/TestCase.php:37]: etc

File: tests/TestCase.php [hunk 4]
```diff
@@ -36,18 +54,14 @@ protected static function getGuzzleMethod()
         return method_exists(ClientInterface::class, 'request') ? 'request' : 'post';
     }
 
-    private static function readObjectAttribute($object, $attributeName)
+    private function phpUnitVersion()
```
Comment (GrahamCampbell) [inline tests/TestCase.php:57]: etc

File: tests/TestCase.php (hunk 2 above)
Comment (imjoehaines) [reply tests/TestCase.php:31]: It can be; I don't think there's a compelling reason for this to be static though?

Comment (GrahamCampbell) [reply tests/TestCase.php:31]: Normally one would write it static unless there is a compelling reason not to.
